### ISSUE TITLE
Unexpected warning "no valid UDP server serving for UDP clients"

### ISSUE THREAD

**Issue by AaronChen0** [shadowsocks/shadowsocks-rust#1632]
When mode `tcp_and_udp` is enabed in shadowsocks-android with no plugin and no udp fallback, the logcat output always contains a warning:
```
WARN main ThreadId(01) shadowsocks_service::local::loadbalancing::ping_balancer: no valid UDP server serving for UDP clients, consider disable UDP with "mode": "tcp_only", currently chose
```
This can be reproduced with the following sslocal config:
```
{
    "locals": [
        {
            "local_address": "127.0.0.1",
            "local_port": 1080,
	    "mode": "tcp_and_udp"
        }
    ],

    "server": "1.1.1.1",
    "server_port": 12345,
    "password": "xxxxxx",
    "method": "aes-256-gcm"
}
```
sslocal with the above config shows that warning. But I can observe udp/quic traffic in my server. Is this warning correct?
```
{
    "locals": [
        {
            "local_address": "127.0.0.1",
            "local_port": 1080,
	    "mode": "tcp_and_udp"
        }
    ],
    "servers": [
        {
            "address": "1.1.1.1",
            "port": 12345,
            "password": "xxxxxx",
            "method": "aes-256-gcm",
        }
    ]
}
```
sslocal with the above config shows no warning and `init chose` both tcp server and udp server. What's the difference between these two configs?

**Comment by zonyitoo** [comment]
There is no `mode` in the root of the configuration object. So `mode` has a default value `tcp_only`.

https://github.com/shadowsocks/shadowsocks-rust/blob/d34e9b72945e6ce3d1cca78cbb97dea6ae7d84a1/crates/shadowsocks-service/src/config.rs#L1522-L1536

For the server instance configuration in BasicFormat, the `mode` of the server instance is the global `mode`:

https://github.com/shadowsocks/shadowsocks-rust/blob/d34e9b72945e6ce3d1cca78cbb97dea6ae7d84a1/crates/shadowsocks-service/src/config.rs#L1911-L1913

So in your first configuration, the server `1.1.1.1:12345` has mode `tcp_only`. But the local instance `127.0.0.1:1080` serves both TCP and UDP SOCKS5 protocol, so when it receives UDP requests, it found no valid server instance that can provide UDP service.

This is what the WARNING log was trying to tell you about.

When there is no valid server could be found, it will automatically fallback to `[0]`, which is the first server instance. So in your case, UDP requests can still be sent to `1.1.1.1:12345` and all went well.

For the second configuration, `servers` is the Extended Format configuration:

https://github.com/shadowsocks/shadowsocks-rust/blob/d34e9b72945e6ce3d1cca78cbb97dea6ae7d84a1/crates/shadowsocks-service/src/config.rs#L2057-L2071

When `mode` is not set in the server instances' own configuration, it will try to use the global `mode` if we are running a Server or Manager. But obviously, we are running a Local server. So the `mode` of the server instance will maintain as it's default `mode`:

https://github.com/shadowsocks/shadowsocks-rust/blob/d34e9b72945e6ce3d1cca78cbb97dea6ae7d84a1/crates/shadowsocks/src/config.rs#L473

which is `tcp_and_udp`.

So that's why you saw the different behavior in your cases.

This should be the expected behavior between Basic Format and Extended Format.

**Comment by AaronChen0** [comment]
Thank you for detailed explanation.
